lib/spack/spack/test/traverse.py

```diff
@@ -284,3 +329,87 @@ def test_breadth_first_versus_depth_first_printing(abstract_specs_chain):
     ^chain-d
 """
     assert s.tree(depth_first=False, cover="edges", **args) == bfs_tree_edges
+
+
+@pytest.fixture()
+def abstract_specs_toposort():
+    # Create a graph that both BFS and DFS would not traverse in topo order, assuming
+    # edges are ordered by target node name. Roots are {A, E} in forward order and
+    # {F, G} in backward order.
+    # forward: DFS([A, E]) traverses [A, B, F, G, C, D, E] (not topo since C < B)
```

Comment:
It is actually an invariant of `traverse_nodes(order="pre/post")` to order edges by target package name.

Reply:
That is good to know: I think this test is much better than I originally thought. However, I don't see any mention of this invariant in the docstring for `traverse_nodes` or in `BaseVisitor.neighbors` (which appears to be the source of this invariant).

Reply:
See also https://github.com/spack/spack/pull/33406#discussion_r1007822909 and the linked PR https://github.com/spack/spack/pull/28673#discussion_r823132555. If I remember correctly, the idea was that this was an invariant thanks to the Spec objects themselves (EdgeMap is always sorted), but in practice even the old traversal code uses an explicit sort, that I didn't touch when refactoring to visitors.

Reply:
I agree that depending on the sorting properties of the edges is sufficient. I think it may be an issue though that we don't express that we use this property to guarantee a fixed, repeatable ordering of edges/nodes. This issue in documentation may predate this PR. It doesn't need to be fixed here.

Reply:
(question) would it also be required to sort `traverse.root_specs` to get a fixed ordering?

Reply:
I don't want to sort them, since there are cases where there's a natural order such as with environment roots

Reply:
This isn't strictly problematic since a user could sort this themselves before passing it in, but complicates the interface since its documentation should mention that the input specs should be pre-sorted to get a fixed ordering. I didn't anticipate a need for users to control traversal orders by sorting the specs themselves (is there a reason for example that users would want the environment spec listing order to influence the traversal order?).